### DISCUSSION TITLE
There are no forward declarations in C#

### DISCUSSION THREAD

**Discussion by MulleDK19** [community/community#7573] [Copilot Conversations]
So, I've been using Copilot exclusively with C#, and one shortcoming is that it has no clue about anything that comes after what you're working on.

So if I have a ThisIsAWeirdlyNamedMethod below Main() and I'm inside main, and I ask it to // Call the weirdly named method
It won't be able to do that, as it has no clue it's there.

Then as a test, I tried some stuff with Go (a language I have no clue on), and noticed the same problem doesn't exist here.

Turns out, in Go, the system is injecting forward declarations into the context, so it at least knows the type and method names that exist, regardless of order.

While forward declarations aren't in C#, putting the signature in a comment seems to work too.

**Comment by Urethramancer** [comment]
It's a bit hit and miss in Go. I've been filling in comments for functions today, and I get three dfferent behaviours at random when starting to type "//" above the functions:
- No suggestion
- A completely bonkers function name and description for the context
- The exact name and an appropriate description

Definitely a feature that needs more work.

**Reply by MulleDK19** [reply]
Like I said, it doesn't know about what comes after the caret. In Go, it only knows that the function exists, not where it is.
If you want it to help you write comments for a function, put the function signature before your comment.

```cpp
//func add(a int, b int) int:
// Start writing your comment here
func add(a int, b int) int
{
    return a + b;
}
```

Then it'll know what function it is documenting.

**Comment by GSPP** [comment]
Can someone from the team comment if this is language-specific?

**Reply by SmallLemur** [reply]
Unfortunately, it is not language specific. Copilot is does not have the forward context to make suggestion at the moment. We are experimenting at the moment to improve the quality of the suggestion

**Reply by MulleDK19** [reply]
Hmm.. pretty sure that last I checked, Copilot adds forwards in Go.

**Reply by Urethramancer** [reply]
I've been enjoying what looks like complete project-awareness  in Go the past couple of weeks. After adding a new function in one file, Copilot starts suggesting I use it in another in (usually) the right spot. Likewise within a single file - I added some help functions below a setup function, and the first suggestions when I added new lines in the top function were what I just added.

**Reply by MulleDK19** [reply]
> I've been enjoying what looks like complete project-awareness in Go the past couple of weeks. After adding a new function in one file, Copilot starts suggesting I use it in another in (usually) the right spot. Likewise within a single file - I added some help functions below a setup function, and the first suggestions when I added new lines in the top function were what I just added.

It's not project aware. What might be happening is the fact that it'll include context of files you have open in the IDE. This applies to all languages.

**Reply by Urethramancer** [reply]
That works out to the same for me. There is other content, and it is aware of it.

**Reply by SmallLemur** [reply]
@MulleDK19 it might be due to some experimentation we are running on a subset of users to have wider awareness.

**Comment by GSPP** [comment]
@SmallLemur I'm kind of familiar with the OpenAI API. I understand you are sending a prefix and GPT continues it.

Could you send the string "[currentDocumentPrefix]###[currentDocumentPostfix]" and train the model on the whole document that way? ### marks the cursor position. That way the whole file would be available as context and the model would still understand where the cursor is.

**Reply by GSPP** [reply]
I find myself rearranging files so that low level functions come before high level functions. I do this just so that Copilot will work (because it's so useful that these contortions actually pay off). It leads to an unnatural order, though, so before checking in, I need to shuffle the file back.

**Comment by thethiny** [comment]
I found this thread after I posted my own, which is the exact same issue I'm facing with Python.
